### DISCUSSION TITLE
Describe a collection of state events as a list (vs. an object)

### DISCUSSION THREAD

**Discussion by cdoublev** [statelyai/xstate#3467] [Q&A] [answered]
Hello,

I started reading the doc seriously 2 days ago, while implementing my own simple (crappy) state machine API to get an idea if the pattern can help me on a current project.

There is [this example in the doc (Guides > Transitions > Wildcard descriptors)](https://xstate.js.org/docs/guides/transitions.html#wildcard-descriptors):

```js
// For SOME_EVENT, the wildcard transition to "elsewhere" will be taken
on: [
  { event: "*", target: "elsewhere" },
  { event: "SOME_EVENT", target: "here" },
]
```

When I try to visualize a state machine (full code below) that contains a list of events associated with a state, the visualizer crashes with this error: `Application error: a client-side exception has occurred (see the browser console for more information).`

Is this example invalid?

<details>
<summary>Full code</summary>

```js
import { createMachine } from 'xstate';

const stepMachine = createMachine({
  id: 'step',
  initial: 'one',
  states: {
    one: {
      on: { NEXT: 'two' }
    },
    two: {
      on: [
        { PREV: 'one' },
        { NEXT: 'three' },
      ],
    },
    three: {
      type: 'final'
    }
  }
});

createMachine(stepMachine)
```
</details>

**Comment by cdoublev** [comment, marked as answer]
Oh sorry, I should stop for today, I think. I will figure out the pros of describing events as a list vs. an object tomorrow, I guess.

<details>
<summary>Correction</summary>

```js

import { createMachine } from 'xstate';

const stepMachine = createMachine({
  id: 'step',
  initial: 'one',
  states: {
    one: {
      on: { NEXT: 'two' }
    },
    two: {
      on: [
        { event: 'PREV', target: 'one' },
        { event: 'NEXT', target: 'three' },
      ],
    },
    three: {
      type: 'final'
    }
  }
});

createMachine(stepMachine)
```
</details>

**Comment by Andarist** [comment]
The list is currently supported but I was thinking about revisiting this in the upcoming major version and potentially removing the support for it. Any particular reason why you like this pattern?

**Comment by cdoublev** [comment]
Nope. I found it intriguing that the list form is not documented elsewhere. I don't have enough experience with XState and FSMs to tell how it could be usefull, sorry! Maybe to apply "conditional eventless" transition?

```js
[
  { condition: isLocked, event: '*' },
  { event: 'BACK', target: 'one' },
  { event: 'NEXT', target: 'three' },
]
```
